File: backend/src/services/openai.service.js
import OpenAI from "openai";
import config from "../config/config.js";
import logger from "../utils/logger.js";
import { AppError } from "../utils/errors.js";

// Initialize OpenAI client
const openai = new OpenAI({
  apiKey: config.openai.apiKey || "dummy-key",
});

/**
 * Call OpenAI Chat Completion with JSON mode enabled.
 *
 * @param {string} systemPrompt - System prompt instructions
 * @param {string} userPrompt - User prompt containing content
 * @param {object} [options] - Additional options (model, temperature)
 * @returns {Promise<object>} Parsed JSON response object
 */
export async function generateJSONCompletion(systemPrompt, userPrompt, options = {}) {
  if (!config.openai.apiKey) {
    throw new AppError(
      "OPENAI_API_KEY is not configured in environment variables.",
      500,
      "OPENAI_SERVICE"
    );
  }

  try {
    const response = await openai.chat.completions.create({
      model: options.model || config.openai.model,
      temperature: options.temperature ?? 0.2,
      response_format: { type: "json_object" },
      messages: [
        { role: "system", content: systemPrompt },
        { role: "user", content: userPrompt },
      ],
    });

    const content = response.choices[0]?.message?.content;
    if (!content) {
      throw new Error("OpenAI returned an empty completion response.");
    }

    return JSON.parse(content);
  } catch (error) {
    logger.error(`OpenAI JSON Completion failed: ${error.message}`);
    throw error;
  }
}

/**
 * Call OpenAI Chat Completion returning plain text (e.g., for cover letters).
 *
 * @param {string} systemPrompt - System prompt instructions
 * @param {string} userPrompt - User prompt
 * @param {object} [options] - Additional options (model, temperature)
 * @returns {Promise<string>} Clean text response
 */
export async function generateTextCompletion(systemPrompt, userPrompt, options = {}) {
  if (!config.openai.apiKey) {
    throw new AppError(
      "OPENAI_API_KEY is not configured in environment variables.",
      500,
      "OPENAI_SERVICE"
    );
  }

  try {
    const response = await openai.chat.completions.create({
      model: options.model || config.openai.model,
      temperature: options.temperature ?? 0.4,
      messages: [
        { role: "system", content: systemPrompt },
        { role: "user", content: userPrompt },
      ],
    });

    const content = response.choices[0]?.message?.content;
    if (!content) {
      throw new Error("OpenAI returned an empty completion response.");
    }

    return content.trim();
  } catch (error) {
    logger.error(`OpenAI Text Completion failed: ${error.message}`);
    throw error;
  }
}

/**
 * Call OpenAI Multimodal Vision Chat Completion with JSON mode enabled to analyze screenshots.
 *
 * @param {string} systemPrompt - System prompt instructions
 * @param {string} userTextPrompt - User text prompt
 * @param {string} base64Data - Base64 image data (with or without data URL prefix)
 * @param {string} [mimeType="image/png"] - Image mime type
 * @param {object} [options] - Additional options
 * @returns {Promise<object>} Parsed JSON response object
 */
export async function generateVisionJSONCompletion(
  systemPrompt,
  userTextPrompt,
  base64Data,
  mimeType = "image/png",
  options = {}
) {
  if (!config.openai.apiKey) {
    throw new AppError(
      "OPENAI_API_KEY is not configured in environment variables.",
      500,
      "OPENAI_SERVICE"
    );
  }

  // Format data URL
  const imageUrl = base64Data.startsWith("data:")
    ? base64Data
    : `data:${mimeType};base64,${base64Data}`;

  try {
    const response = await openai.chat.completions.create({
      model: options.model || config.openai.model,
      temperature: options.temperature ?? 0.1,
      response_format: { type: "json_object" },
      messages: [
        { role: "system", content: systemPrompt },
        {
          role: "user",
          content: [
            { type: "text", text: userTextPrompt },
            {
              type: "image_url",
              image_url: {
                url: imageUrl,
                detail: "high",
              },
            },
          ],
        },
      ],
    });

    const content = response.choices[0]?.message?.content;
    if (!content) {
      throw new Error("OpenAI returned an empty vision completion response.");
    }

    return JSON.parse(content);
  } catch (error) {
    logger.error(`OpenAI Vision JSON Completion failed: ${error.message}`);
    throw error;
  }
}

export default {
  generateJSONCompletion,
  generateTextCompletion,
  generateVisionJSONCompletion,
};

File: resume-api/services/openai_service.py
import json
import logging
import re
from typing import Any, Dict, Optional

from config import settings
from openai import OpenAI
from prompts import format_job_extraction_prompt, job_extraction_system_prompt

logger = logging.getLogger(__name__)

_openai_client = None


def get_openai_client() -> Optional[OpenAI]:
    global _openai_client
    if _openai_client is None:
        if settings.OPENAI_API_KEY:
            _openai_client = OpenAI(api_key=settings.OPENAI_API_KEY)
        else:
            logger.warning("OPENAI_API_KEY not configured. LLM calls will fail or use fallback.")
            _openai_client = None
    return _openai_client


def clean_and_parse_json(raw_text: str, default: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    """
    Robustly extract and parse JSON from raw LLM output strings,
    stripping markdown code fences and extraneous text.
    """
    if default is None:
        default = {}

    if not raw_text or not isinstance(raw_text, str):
        return default

    cleaned = raw_text.strip()

    # Strip markdown code blocks if present
    if cleaned.startswith("```"):
        cleaned = re.sub(r"^```(?:json)?\s*", "", cleaned, flags=re.IGNORECASE)
        cleaned = re.sub(r"\s*```$", "", cleaned)
        cleaned = cleaned.strip()

    # Find boundaries of JSON object or array
    start_brace = cleaned.find("{")
    start_bracket = cleaned.find("[")

    if start_brace != -1 and (start_bracket == -1 or start_brace < start_bracket):
        end_brace = cleaned.rfind("}")
        if end_brace != -1 and end_brace > start_brace:
            cleaned = cleaned[start_brace : end_brace + 1]
    elif start_bracket != -1:
        end_bracket = cleaned.rfind("]")
        if end_bracket != -1 and end_bracket > start_bracket:
            cleaned = cleaned[start_bracket : end_bracket + 1]

    try:
        return json.loads(cleaned)
    except json.JSONDecodeError:
        # Fallback: remove trailing commas before closing braces/brackets
        try:
            fixed = re.sub(r",\s*([}\]])", r"\1", cleaned)
            return json.loads(fixed)
        except Exception as err:
            logger.error("Failed to parse JSON output from LLM: %s. Snippet: %s", err, raw_text[:200])
            return default


def tailor_resume_with_openai(
    job_description: str,
    master_resume_data: Dict[str, Any],
) -> Dict[str, Any]:
    """
    Calls OpenAI API using ApplyPilot's ATS 92+ prompt to extract job info
    and tailor the candidate profile specifically to the job description.

    Returns:
        Dict containing tailored 'candidate' and 'job' objects.
    """
    client = get_openai_client()
    if not client:
        logger.warning("No OpenAI client available. Returning master resume data.")
        return {
            "candidate": master_resume_data,
            "job": {"details": {"title": "Target Role", "description": job_description[:200]}},
        }

    system_prompt = job_extraction_system_prompt
    user_prompt = format_job_extraction_prompt(
        html_content=job_description,
        resume_content=json.dumps(master_resume_data, indent=2),
    )

    try:
        messages = [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt},
        ]

        response = client.chat.completions.create(
            model=settings.OPENAI_MODEL,
            messages=messages,
            temperature=0.7,
            response_format={"type": "json_object"},
        )

        raw_content = response.choices[0].message.content or "{}"
        parsed = clean_and_parse_json(raw_content)

        candidate = parsed.get("candidate")
        if not candidate or not isinstance(candidate, dict):
            logger.warning("OpenAI response did not contain 'candidate' object, using fallback.")
            parsed["candidate"] = master_resume_data

        return parsed

    except Exception as e:
        logger.error("OpenAI resume tailoring call failed: %s", e, exc_info=True)
        return {
            "candidate": master_resume_data,
            "job": {"details": {"title": "Target Role", "description": job_description[:200]}},
        }
